src/main/java/com/snowplowanalytics/snowplow/tracker/emitter/EventStore.java

```diff
@@ -0,0 +1,20 @@
+package com.snowplowanalytics.snowplow.tracker.emitter;
+
+import java.util.Collection;
+import java.util.List;
+
+import com.snowplowanalytics.snowplow.tracker.payload.TrackerEvent;
+
+public interface EventStore {
+
+    boolean add(TrackerEvent trackerEvent);
+
+    void removeAllEvents(List<TrackerEvent> eventsList);
+
+    // this can be removed if Emitter.getBuffer is changed to return just the size
+    List<TrackerEvent> retrieveAllEvents();
+
+    void prepareAllEventsForRemoval();
```

Comment:
I agree it's confusing. I was trying to keep the current BatchEmitter functionality, but I don't know how useful it is. There are two methods I had trouble with:

1. `BatchEmitter.flushBuffer()`. This method moved all events from eventBuffer into the stagingBuffer, using `poll`/`offer`, then tried to send them all without regard for batch size. So there are two aspects - the batching ignoring, and the queue management part. For the latter, I created `prepareAllEventsForRemoval()`. I think we should simplify this method and just try to send everything from stagingBuffer queue - or as you mention below, use only one queue anyway.

2. `BatchEmitter.getBuffer()` copies all the events from stagingBuffer into an arraylist. The events are not removed from the queue, so it's not the same as `storage.removeAllEvents()`. This is why I made `retrieveAllEvents()`. Is `getBuffer()` really worthwhile? It's used in one of the BatchEmitter tests.

Reply:
Having `flushBuffer()` send all the events in one go seems like it would create oversized payloads?

Reply:
Your point 1 is interesting. Honestly, I've never checked but there is the risk that batching is screwed up by that. We may consider `flushBuffer` as a sort of emergency exit so I think it would be ok anyway. Good point btw!

Point 2: `getBuffer` is public interface of Emitter. However, digging in the past of that method it seems to exist as method needed just for the test. I would say... let's keep `retrieveAllEvents()` in the EventStore as you correctly did. Then we can think to remove `getBuffer` from Emitter in a future task.